catgrad-core/src/interpreter/backend/candle.rs

```diff
@@ -0,0 +1,636 @@
+        if x_shape != y_shape {
+            // Check if one shape can be squeezed to match the other
+            if x_shape.len() > y_shape.len() && x_shape.ends_with(&[1]) {
+                // x has extra dimension of size 1, squeeze it
+                let x_squeezed = x.0.squeeze(x_shape.len() - 1).unwrap();
+                CandleTensor(op(x_squeezed, y.0).unwrap())
+            } else if y_shape.len() > x_shape.len() && y_shape.ends_with(&[1]) {
+                // y has extra dimension of size 1, squeeze it
+                let y_squeezed = y.0.squeeze(y_shape.len() - 1).unwrap();
+                CandleTensor(op(x.0, y_squeezed).unwrap())
+            } else if y_shape.len() == 1 && y_shape[0] == 1 {
+                // y is a scalar, broadcast it
+                let y_broadcast = y.0.broadcast_as(x_shape).unwrap();
+                CandleTensor(op(x.0, y_broadcast).unwrap())
+            } else if x_shape.len() == 1 && x_shape[0] == 1 {
+                // x is a scalar, broadcast it
+                let x_broadcast = x.0.broadcast_as(y_shape).unwrap();
+                CandleTensor(op(x_broadcast, y.0).unwrap())
+            } else {
+                // Try the operation directly
+                CandleTensor(op(x.0, y.0).unwrap())
+            }
+        } else {
+            CandleTensor(op(x.0, y.0).unwrap())
+        }
+    }
+
+    // ============================================================================
+    // TYPE-SPECIFIC OPERATION FUNCTIONS
+    // ============================================================================
+    //
+    // WHY WE NEED SEPARATE add_f32/add_u32 FUNCTIONS (vs ndarray's generic approach):
+    //
+    // The ndarray backend can use generic functions like:
+    //   fn add<D>(x: ArrayD<D>, y: ArrayD<D>) -> ArrayD<D> { x + y }
+    //
+    // But Candle requires separate functions because:
+    //
+    // 1. **Type Erasure**: CandleTensor wraps `candle_core::Tensor` which uses runtime
+    //    type dispatch via `DType` enum, not Rust generics. The tensor type is not
+    //    generic over element types like `ArrayD<D>`.
+    //
+    // 2. **API Design**: `candle_core::Tensor` operations are not generic - they work
+    //    on the concrete `Tensor` type and dispatch internally based on `DType`.
+    //
+    // 3. **Pattern Matching**: The main Backend trait methods need to match on
+    //    TaggedNdArrayTuple variants (F32/U32) and call the appropriate function.
+    //
+    // 4. **Consistency**: Even though both functions do the same thing (x.0 + y.0),
+    //    having separate functions makes the code more explicit and matches the
+    //    pattern used throughout the Candle backend.
+    //
+    // This is a fundamental difference in how ndarray (generic) vs Candle (type-erased)
+    // handle multi-type tensor operations.
+    // ============================================================================
+
+    fn add_f32(x: CandleTensor, y: CandleTensor) -> CandleTensor {
```

Comment:
Catgrad ops should never do implicit broadcasting: this is a specific design choice which is not currently very clear in the docs for `Backend`.

Fix:

- add_u32/add_f32 helpers can be removed, and just write a+b inline in the `add` kernel.
- same for other similar ops (mul, div, etc.)
- For now, if two tensors do not have compatible shapes, kernels should panic.

Created issue https://github.com/hellas-ai/catgrad/issues/189 to track.